lib/string_map.h

```diff
@@ -0,0 +1,226 @@
+
+    bool empty() const noexcept { return data.empty(); }
+    size_type size() const noexcept { return data_map.size(); }
+    size_type max_size() const noexcept { return data_map.max_size(); }
+    bool operator==(const string_map &a) const { return data == a.data; }
+    bool operator!=(const string_map &a) const { return data != a.data; }
+    void clear() {
+        data.clear();
+        data_map.clear();
+    }
+
+    // Functions above do have `std::string_view` versions. Here we are having
+    // important special case: if `a` is not something that was interned, we do
+    // not copy / intern it, we know for sure that `a` is not in the map and we
+    // do not need to perform a lookup
+    iterator find(cstring a) { return tr_iter(data_map.find(a)); }
+    iterator find(std::string_view a) {
+        cstring key = cstring::get_cached(a);
+        if (key.isNull()) return data.end();
+
+        return tr_iter(data_map.find(key));
+    }
+    const_iterator find(cstring a) const { return tr_iter(data_map.find(a)); }
+    const_iterator find(std::string_view a) const {
+        cstring key = cstring::get_cached(a);
+        if (key.isNull()) return data.end();
+
+        return tr_iter(data_map.find(key));
+    }
+
+    size_type count(cstring a) const { return data_map.count(a); }
+    size_type count(std::string_view a) const {
+        cstring key = cstring::get_cached(a);
+        if (key.isNull()) return 0;
+
+        return data_map.count(key);
+    }
+
+    bool contains(cstring a) const { return data_map.contains(a); }
+    bool contains(std::string_view a) const {
+        cstring key = cstring::get_cached(a);
+        if (key.isNull()) return false;
+
+        return data_map.contains(key);
+    }
+
+    template <typename Key>
+    V &operator[](Key &&k) {
+        auto it = find(key_type(std::forward<Key>(k)));
+        if (it == data.end()) {
+            it = data.emplace(data.end(), k, V());
+            data_map.emplace(it->first, it);
+        }
+        return it->second;
+    }
+
+    template <typename Key>
+    V &at(Key &&k) {
+        auto it = find(std::forward<Key>(k));
+        if (it == data.end()) throw std::out_of_range("string_map::at");
+        return it->second;
+    }
+    template <typename Key>
+    const V &at(Key &&k) const {
+        auto it = find(std::forward<Key>(k));
+        if (it == data.end()) throw std::out_of_range("string_map::at");
+        return it->second;
+    }
+
+    template <typename Key, typename... VV>
+    std::pair<iterator, bool> emplace(Key &&k, VV &&...v) {
+        auto it = find(key_type(std::forward<Key>(k)));
+        if (it == data.end()) {
+            it = data.emplace(data.end(), std::piecewise_construct_t(), std::forward_as_tuple(k),
+                              std::forward_as_tuple(std::forward<VV>(v)...));
+            data_map.emplace(it->first, it);
+            return std::make_pair(it, true);
```

Comment:
just `std::pair` (or maybe even `{it, true}`) should be sufficent.

Reply:
Yeah. It's just muscle memory :)